lib/tasks/security.rake

```diff
@@ -1,7 +1,29 @@
+require "bcrypt"
+
 namespace :security do
   desc "Import the NCSC's password denylist.  This clears the current denylist."
   task import_ncsc_denylist: :environment do
     count = BannedPassword.import_from_ncsc
     puts "imported #{count} passwords"
   end
+
+  desc "Report on usage of passwords on NCSC's password denylist."
+  task report_ncsc_denylist_usage: :environment do
+    User.where(ncsc_password_match: nil).each do |user|
+      user_hashed_password = user.encrypted_password # pragma: allowlist secrets
+      bcrypt = ::BCrypt::Password.new(user_hashed_password)
+
+      hashed_nscs_password_list = BannedPassword.pluck(:password).map do |password|
+        if User.pepper.present?
+          password = "#{password}#{User.pepper}" # pragma: allowlist secrets
+        end
+        ::BCrypt::Engine.hash_secret(password, bcrypt.salt)
+      end
+
+      user.update!(ncsc_password_match: hashed_nscs_password_list.include?(user_hashed_password))
```

Comment:
This could be fairly slow as we're effectively constructing a rainbow table for each user and then checking if we get a hit.

I think this might be a bit faster, as it likely won't need to hash *all* the banned passwords if there is a hit:

```ruby
has_banned_password = BannedPassword.pluck(:password).any? do |password|
  User. valid_password?(password)
end

user.update!(ncsc_password_match: has_banned_password)
```

It also means we don't need to know the details of how devise hashes passwords.

Reply:
Thanks!  This does seem considerably faster and has simplified the code.